listenbrainz/spotify_updater/spotify_read_listens.py

```diff
@@ -0,0 +1,246 @@
+            break
+        except (InternalServerError, ServiceUnavailable) as e:
+            retries -= 1
+            current_app.logger.info('ISE while trying to import listens for %s: %s', str(user), str(e))
+            if retries == 0:
+                raise spotify.SpotifyListenBrainzError('ISE while trying to import listens: %s', str(e))
+
+    # we've succeeded so update the last_updated field for this user
+    spotify.update_last_updated(user.user_id)
+
+
+def process_all_spotify_users():
+    """ Get a batch of users to be processed and import their Spotify plays.
+
+    Returns:
+        (success, failure) where
+            success: the number of users whose plays were successfully imported.
+            failure: the number of users for whom we faced errors while importing.
+    """
+    current_app.logger.info('Getting list of users to be processed...')
+    try:
+        users = spotify.get_active_users_to_process()
+    except DatabaseException as e:
+        current_app.logger.error('Cannot get list of users due to error %s', str(e), exc_info=True)
+        return 0, 0
+
+    if not users:
+        return 0, 0
+
+    success = 0
+    failure = 0
+    for u in users:
+        t = time.time()
+        current_app.logger.info('Importing spotify listens for user %s', str(u))
+        try:
+            process_one_user(u)
+            success += 1
+        except spotify.SpotifyAPIError as e:
+            # if it is an error from the Spotify API, show the error message to the user
+            spotify.update_last_updated(
+                user_id=u.user_id,
+                success=False,
+                error_message=str(e),
+            )
+            failure += 1
+        except spotify.SpotifyListenBrainzError as e:
+            current_app.logger.critical('spotify_reader could not import listens: %s', str(e), exc_info=True)
+            failure += 1
+        except Exception as e:
+            current_app.logger.critical('spotify_reader could not import listens: %s', str(e), exc_info=True)
+            failure += 1
+
+        current_app.logger.info('Took a total of %.2f seconds to process user %s', time.time() - t, str(u))
+
+    current_app.logger.info('Processed %d users successfully!', success)
+    current_app.logger.info('Encountered errors while processing %d users.', failure)
+    return success, failure
+
+
+def main():
+    app = listenbrainz.webserver.create_app()
+    with app.app_context():
+        current_app.logger.info('Spotify Reader started...')
+        while True:
+            t = time.time()
+            success, failure = process_all_spotify_users()
+            if success + failure > 0:
+                current_app.logger.info('All %d users in batch have been processed.', success + failure)
+                current_app.logger.info('Total time taken: %.2f s, average time per user: %.2f s.', time.time() - t, (time.time() - t) / (success + failure))
+            time.sleep(10)
```

Comment:
I wonder if this should be dynamic -- 10 seconds should work fine for now, but over time as more people use our stuff 10 seconds is time wasted we may need to use. 

For instance, if the last update run didn't fetch anything new, we can wait longer. But if we got a pile of updates, we may not want to wait 10 seconds.

Reply:
[LB-376](https://tickets.metabrainz.org/browse/LB-376)